src/v/raft/log_eviction_stm.h

```diff
@@ -56,18 +67,32 @@ class log_eviction_stm {
     }
 
 private:
+    enum class eviction_event_type { storage = 0, kafka };
+    struct eviction_event {
+        model::offset offset;
+        eviction_event_type event_type;
+    };
+    static ss::sstring eviction_event_type_to_str(eviction_event_type);
+
     void monitor_log_eviction();
     void process_log_eviction();
     ss::future<> housekeeping();
-    ss::future<> do_process_log_eviction(model::offset);
+    ss::future<model::offset> do_process_log_eviction(eviction_event);
+
+    ss::future<> apply(model::record_batch) override;
+    ss::future<> handle_eviction() override { return ss::now(); }
```

Comment:
Good point. However is that necessary? This is the eviction STM, so it drives eviction events, could there ever be a case where `handle_eviction` is called?

Reply:
Yeah, imagine a case where an out of date node re-joins the cluster (perhaps after a long-ish connectivity issue). The STM was still spinning in that case. I think calling advancing `_next` here to the start offset is fine here, since the the special eviction batch must be replicated on all nodes. This means that we can't miss one such batch by advancing (if one node was isloated, replicating the batch would have failed).

I'm not sure about wheter we should advance to raft's start offset or raft's start offset plus the start offset delta added in this PR.

Reply:
`_next` can fall bellow the start offset if the node becomes isolated and re-joins later on. Local retention does not replicate a `prefix_truncate` batch, so I think that after a period of isolation where local retention kicked in, `state_machine::apply` will try to create a reader starting from offset `_next`. Since local retention moved the start offset forward and `_next` hasn't been updated as no batches were applied, `_next` will now be below the start offset.

Shouldn't be too hard to try out this scenario and see what happens.